cpp/lib/include/Xerces.h

```diff
@@ -0,0 +1,123 @@
+        std::string chars_;
+        inline virtual Type getType() const override { return CHARACTERS; }
+    };
+
+    struct IgnorableWhitespace : Characters {
+        inline virtual Type getType() const override { return IGNORABLE_WHITESPACE; }
+    };
+
+private:
+    template<typename XmlPartType> static const std::shared_ptr<const XmlPartType> CastToXmlPartOrDie(const std::shared_ptr<const XmlPart> part, const XmlPart::Type part_type) {
+        if (unlikely(part->getType() != part_type))
+            LOG_ERROR("Could not convert XmlPart to " + XmlPart::TypeToString(part_type));
+        return std::static_pointer_cast<const XmlPartType>(part);
+    }
+public:
+    static const std::shared_ptr<const StartElement> CastToStartElementOrDie(const std::shared_ptr<const XmlPart> part) { return CastToXmlPartOrDie<StartElement>(part, XmlPart::START_ELEMENT); }
+    static const std::shared_ptr<const EndElement> CastToEndElementOrDie(const std::shared_ptr<const XmlPart> part) { return CastToXmlPartOrDie<EndElement>(part, XmlPart::END_ELEMENT); }
+    static const std::shared_ptr<const ProcessingInstruction> CastToProcessingInstructionOrDie(const std::shared_ptr<const XmlPart> part) { return CastToXmlPartOrDie<ProcessingInstruction>(part, XmlPart::PROCESSING_INSTRUCTION); }
+    static const std::shared_ptr<const Characters> CastToCharactersOrDie(const std::shared_ptr<const XmlPart> part) { return CastToXmlPartOrDie<Characters>(part, XmlPart::CHARACTERS); }
+    static const std::shared_ptr<const IgnorableWhitespace> CastToIgnorableWhitespaceOrDie(const std::shared_ptr<const XmlPart> part) { return CastToXmlPartOrDie<IgnorableWhitespace>(part, XmlPart::IGNORABLE_WHITESPACE); }
+private:
+    class Handler : public xercesc::HandlerBase {
+        friend class Xerces;
+        Xerces * xerces_;
+
+    public:
+        void characters(const XMLCh * const chars, const XMLSize_t length);
+        void endElement(const XMLCh * const name);
+        void ignorableWhitespace(const XMLCh * const chars, const XMLSize_t length);
+        void processingInstruction(const XMLCh * const target, const XMLCh *const data);
+        void startElement(const XMLCh * const name, xercesc::AttributeList &attributes);
+    };
+
+    class ErrorHandler : public xercesc::ErrorHandler {
+    public:
+        void warning(const xercesc::SAXParseException &exc) { LOG_WARNING(Xerces::ToString(exc.getMessage())); }
+        void error(const xercesc::SAXParseException &exc) { LOG_WARNING(Xerces::ToString(exc.getMessage())); }
+        void fatalError(const xercesc::SAXParseException &exc) { LOG_ERROR(Xerces::ToString(exc.getMessage())); }
+        void resetErrors() {}
+    };
+
+    Handler* handler_;
+    ErrorHandler* error_handler_;
+    std::queue<std::shared_ptr<XmlPart>> buffer_;
+    void addToBuffer(std::shared_ptr<XmlPart> xml_part) { buffer_.push(xml_part); }
+    friend class Handler;
+public:
+    Xerces(const std::string &xml_file);
```

Comment:
I would prefer adding "explicit" here.  Also, not sure if we should really call this Xerces.  That is extremely cryptic and leaks an implementation detail!  How about "XMLParser"?